### PR TITLE
[WIP] Adds support to submission worker for updating challenge fields #840 

### PR DESCRIPTION
**Pull request by yadavankit**
@taranjeet @deshraj please review and add suggestions :) #840

### CODE REVIEW

**Comment by yadavankit** [comment]
@taranjeet can you please see if am approaching this in correct way? Also, please suggest changes :)

**Comment by taranjeet** [comment]
@yadavankit : been busy with office these days. Will get back to you positively by tomorrow. Thanks for your time and help :)

File: scripts/workers/submission_worker.py [hunk 1]
```diff
@@ -483,6 +500,9 @@ def main():
     add_challenge_queue_name = '{hostname}_{process_id}'.format(hostname=socket.gethostname(),
                                                                 process_id=str(os.getpid()))
 
+    edit_challenge_queue_name = '{hostname}_{process_id}'.format(hostname=socket.gethostname(),
```
Comment (taranjeet) [inline scripts/workers/submission_worker.py]: We have planned to use the same queue, but different routing keys.

File: apps/challenges/models.py [hunk 2]
```diff
@@ -80,6 +84,13 @@ def is_active(self):
         return False
 
 
+@receiver(post_save, sender=Challenge)
+def publish_message_on_change(instance, created=False, **kwargs):
```
Comment (taranjeet) [inline apps/challenges/models.py]: I think that this PR is missing a very important thing. This signal handler will be called when every any field of model instance is updated and corresponding to it, a message will be published each time. If we clearly see that the worker only needs challenge id (which will never change) to maintain map, and then corresponding files, we come to a realization that the worker should only be notified of changes in file field. 

This way we prevent ourselves and the worker from unnecessarily loading files again and again.

File: scripts/workers/submission_worker.py [hunk 3]
```diff
@@ -461,6 +467,17 @@ def add_challenge_callback(ch, method, properties, body):
         traceback.print_exc()
 
 
+def edit_challenge_callback(ch, method, properties, body):
+    try:
+        logger.info("[x] Received edit challenge message %s" % body)
+        body = yaml.safe_load(body)
+        process_edit_challenge_message(body)
```
Comment (taranjeet) [inline scripts/workers/submission_worker.py]: As far as the issue is concerned, I think something is missing here also. The file field which have changes needs to be downloaded again and then imported. I think here the code to download those files is missing. 

I may be wrong here, let me know if something is not clear here.

**Comment by taranjeet** [comment]
@yadavankit : Sorry for the delay in reviewing. I have left some major concept level feedback. Let me know if I missed something or if anything is not clear to you.

Thanks for your help.

File: scripts/workers/submission_worker.py (hunk 1 above)
Comment (yadavankit) [inline scripts/workers/submission_worker.py]: Sure, will be routing this with different routing keys and same queue.

File: apps/challenges/models.py (hunk 2 above)
Comment (yadavankit) [inline apps/challenges/models.py]: Yeah, I actually wanted to get rid of this too. `post_save` signal provides `update_fields` and that could have done the task, but for some reasons, `update_fields` is `None` when `post_save` signal is invoked. Am looking into that and will follow up asap.

Comment (yadavankit) [inline apps/challenges/models.py]: @taranjeet sorry for the delay, can you please help me figure out why `update_fields` is coming `None`. For now, I am trying [this](https://stackoverflow.com/questions/36719566/identify-the-changed-fields-in-django-post-save-signal) solution. Please see if its the correct approach.

**Comment by RishabhJain2018** [comment]
@yadavankit Please update the status of this PR? Is it ready to be merged ?

**Comment by yadavankit** [comment]
@RishabhJain2018 no its not actually, I have been very busy with my Startup's work for the last two months. Am sorry for the delay caused in this, but I couldn't help it.

**Comment by deshraj** [comment]
Thanks for letting us know @yadavankit. 

Closing this since not complete.

**Comment by isht3** [comment]
@yadavankit Can I take over from your progress in this PR? If yes, could you tell me what all should be should be added to this? :smile:

**Comment by yadavankit** [comment]
@isht3 Apologies, but actually am not an active contributor here. Thanks to loads of work at my startup :P You can have a look at the code and comments by @taranjeet for starters

**Comment by isht3** [comment]
@yadavankit That's cool! All the best and thanks!

**Pull request closed without merging**